### DISCUSSION TITLE
nvim_set_hl color remove problem

### DISCUSSION THREAD

**Discussion by EdiDD** [neovim/neovim#24405] [Q&A] [answered]
Hi,
when i set in nvim 0.9.1:
`api.nvim_set_hl(0, 'TSFunction', { bold=true, italic=true })` it will remove default color and set black instead. How to preserve default colors for fg and bg ?

**Comment by clason** [comment]
You can't. Use the get_hl API to get the current properties, modify them, and then set these.

**Comment by EdiDD** [comment]
Thanks for help. Could you share some example ?

**Comment by seandewar** [comment, marked as answer]
You can create a utility function, especially if you want to do this a lot:

```lua
function modify_hl(ns, name, changes)
  local def = vim.api.nvim_get_hl(ns, { name = name, link = false })
  vim.api.nvim_set_hl(ns, name, vim.tbl_deep_extend("force", def, changes))
end
```

Then in your case: `modify_hl(0, "TSFunction", { bold = true, italic = true })`

_Note:_ `modify_hl` assumes that if the group is linked, that you want to resolve the "effective" definition and modify that.
If you instead want to set fallback attributes in case the group is unlinked later (unlikely), remove `link = false`.

**Comment by EdiDD** [comment]
@seandewar thanks for your answer. Is calling `api.nvim_set_hl` and changing many hl better then calling `vim.cmd([[ .... ]])`  ?

**Comment by seandewar** [comment]
Depends what you mean by "better".
Using `:highlight` can potentially be more convenient, using the API may allow more control.

For example, as in my previous answer, the API is easily able to resolve the effective definition of a linked group and modify its attributes (rather than just setting fall-back attributes in case it's unlinked later, which is what `:highlight` would do).

There are also other cases where the API is clearly preferable. E.g: when manipulating highlights in a specific namespace or when using group IDs.

If performance is your concern, then you're likely over-thinking; there's unlikely any significant difference for just a few groups.
In general, the API avoids string parsing and a trip through the Vim script subsystem, but you should always measure first.
